services/provider/server/consumer.go

```diff
@@ -0,0 +1,175 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	ocsv1alpha1 "github.com/red-hat-storage/ocs-operator/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ocsConsumerManager struct {
+	client       client.Client
+	namespace    string
+	nameByTicket map[string]string
+	nameByUID    map[types.UID]string
+	mutex        sync.RWMutex
+}
+
+func newConsumerManager(ctx context.Context, client client.Client, namespace string) (*ocsConsumerManager, error) {
+	consumers := &ocsv1alpha1.StorageConsumerList{}
+	err := client.List(ctx, consumers)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list storage consumers. %v", err)
+	}
+
+	nameByTicket := map[string]string{}
+	nameByUID := map[types.UID]string{}
+
+	for _, consumer := range consumers.Items {
+		nameByUID[consumer.UID] = consumer.Name
+
+		if ticket, ok := consumer.GetAnnotations()[TicketAnnotation]; ok {
+			nameByTicket[ticket] = consumer.Name
+		}
+	}
+
+	return &ocsConsumerManager{
+		client:       client,
+		namespace:    namespace,
+		nameByTicket: nameByTicket,
+		nameByUID:    nameByUID,
+	}, nil
+}
+
+// Create creates a new storageConsumer resource, updates the consumer cache and returns the storageConsumer UID
+func (c *ocsConsumerManager) Create(ctx context.Context, name, ticket string, capacity resource.Quantity) (string, error) {
+	c.mutex.RLock()
+	if _, ok := c.nameByTicket[ticket]; ok {
+		c.mutex.RUnlock()
+		return "", fmt.Errorf("onboarding ticket is already being used by another storageConsumer")
```

Comment:
Can we set appropriate error code here which is `codes.AlreadyExists`?

Reply:
@iamniting This is the consumer manager code, not the server code. 
I don't think we should introduce JRPC codes here

Reply:
In that case we should read this error and set the grpc code where we are calling this func.

Reply:
Sending `AlreadyExists` status code when onboarding ticket is already being used.